chore: update eslint-config-eslint exports

mdjermanovic opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

Refs discussion in https://github.com/eslint/espree/pull/577#discussion_r1249925279.

Currently, eslint-config-eslint exports a single config array suitable for CommonJS projects. That's fine for this repo, but most of the other repos are now ESM. Usage in an ESM repo would look like in https://github.com/eslint/espree/pull/577:

```js
import eslintConfigEslint from "eslint-config-eslint";
import nodeRecommendedModule from "eslint-plugin-n/configs/recommended-module.js";
import nodeRecommendedScript from "eslint-plugin-n/configs/recommended-script.js";

export default [
    // eslintConfigEslint[3] is eslint-plugin-n's recommended-script config
    ...eslintConfigEslint.slice(0, 3),
    {
        files: ["**/*.js"],
        ...nodeRecommendedModule
    },
    {
        files: ["**/*.cjs"],
        ...nodeRecommendedScript
    },
    ...eslintConfigEslint.slice(4)
];
```

That doesn't look very convenient.

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This PR changes eslint-config-eslint to export an object with 3 configs: `base` (config array), `commonjs` (config object), `esm` (config object), so that the usage would look like this:

```js
import eslintConfigEslint from "eslint-config-eslint";

export default [
    ...eslintConfigEslint.base,
    {
        files: ["**/*.js"],
        ...eslintConfigEslint.esm
    },
    {
        files: ["**/*.cjs"],
        ...eslintConfigEslint.commonjs
    }
];
```

#### Is there anything you'd like reviewers to focus on?

Do the names make sense?

mdjermanovic commented on packages/eslint-config-eslint/index.js:
```diff
@@ -13,17 +14,28 @@ const unicorn = require("eslint-plugin-unicorn");
     rules: {
-        "n/callback-return": ["error", ["cb", "callback", "next"]],
-        "n/handle-callback-err": ["error", "err"],
-        "n/no-deprecated-api": "error",
```
I removed `n/no-deprecated-api` from here because I believe it's already in eslint-plugin-n's recommended configs. Is that correct?

aladdin-add replied:
👍

nzakas commented:
Rather than forcing us to mix and match individual configs, what if we just exported an ESM-compatible array by default and included a CommonJS one under `eslint-config-eslint/cjs`?

That way, the default is equivalent to what we have now and the special case of CommonJS is set off to the side. 

So you'd be able to do this:

```
import eslintConfigESLint from "eslint-config-eslint";

export default [
    ...eslintConfigESLint
];
```

And then in a CommonJS file do:

```
const eslintConfigESLint = require("eslint-config-eslint/cjs");

module.exports = [
    ...eslintConfigESLint
];
```

mdjermanovic commented on packages/eslint-config-eslint/index.js:
```diff
@@ -13,17 +14,28 @@ const unicorn = require("eslint-plugin-unicorn");
-        "n/no-deprecated-api": "error",
+        ...nodeRecommendedScriptConfig.rules,
+        ...nodeSharedRules,
         "n/no-mixed-requires": "error",
         "n/no-new-require": "error",
         "n/no-path-concat": "error"
```
These 3 rules will be in the `commonjs` config only because I believe they don't apply to ESM. Is that correct?

mdjermanovic commented:
> Rather than forcing us to mix and match individual configs, what if we just exported an ESM-compatible array by default and included a CommonJS one under `eslint-config-eslint/cjs`?
> 
> That way, the default is equivalent to what we have now and the special case of CommonJS is set off to the side.

That's another option, but I think the usage would be more complicated in our ESM projects, where we have `.js` (ESM) and `.cjs` files. It would be something like this:

```js
import eslintConfigESLint from "eslint-config-eslint";
import eslintConfigESLintCJS from "eslint-config-eslint/cjs";

export default [
    ...eslintConfigESLint.map(config => ({
        files: ["**/*.js"],
        ...config
    })),
    ...eslintConfigESLintCJS.map(config => ({
        files: ["**/*.cjs"],
        ...config
    }))
];
```

nzakas commented:
To me that seems less complicated because you don't have to mix and match a base with specific ESM or CJS configs. I'd like to plan for a future without CommonJS and design a solution that works optimally in that world even if its suboptimal in a current CommonJS repo.

mdjermanovic commented:
> To me that seems less complicated because you don't have to mix and match a base with specific ESM or CJS configs. I'd like to plan for a future without CommonJS and design a solution that works optimally in that world even if its suboptimal in a current CommonJS repo.

Makes sense, I'll update the exports.

mdjermanovic commented:
I've updated exports so that the default export is the full configuration for ESM files, and there is an additional export `/cjs` that is the full configuration for CJS files.

I also added `package.exports` and updated `package.engine` because some plugins require `"node": ">=16.0.0"`.

mdjermanovic commented:
The problem with `files` in shareable configs occurs when a shareable config should only be applied to some files. For example, in the `eslint.org` repo we have tools that run in Node (currently CommonJS modules but could be converted to ESM), modules that will be bundled in the Playground, and browser scripts. Assuming that we'll eventually convert tools to ESM, eslint-config-eslint's default export (or the `esm` export in your version) config array looks like a good candidate for those files, but not for the Playground modules and browser scripts, so it should be applied to `tools/` only. If the config array contains a mix of ESM and CJS config items, restricting the config to a directory becomes complicated. I don't have a strong opinion on this though, and this is certainly a good exercise on creating and using shareable flat configs.

aladdin-add commented:
well, I see your point!

in my implementation, it has exported 3 configs: `base`, `esm`, `cjs`. They can easily be used directly in specific types of apps.

`esm/cjs`: only used for node.js apps (most eslint's repos should be using it).
`base`: for web apps like eslint.org, a new config can be created based on `base`.

I am happy with the current implementation, and I am somewhat discovering the flexibility of the new config system, we may need a little time to explore the best way to use it. 😄

aladdin-add commented on packages/eslint-config-eslint/package.json:
```diff
@@ -3,15 +3,22 @@
   "version": "8.0.0",
   "author": "Nicholas C. Zakas <nicholas+npm@nczconsulting.com>",
   "description": "Default ESLint configuration for ESLint projects.",
+  "exports": {
+    "./package.json": "./package.json",
+    ".": "./index.js",
+    "./cjs": "./cjs.js"
+  },
```
I didn't see a need to add this - we did this in the eslint package to avoid other devs using the internal modules, but this package is (mostly) only used by eslint. :)

mdjermanovic replied:
I thought it would be good to clarify which modules are intended to be public.

And with this, `import "eslint-config-eslint/cjs"` doesn't need `.js` extension :)

aladdin-add replied:
we need to export `./eslintrc` as well - it was used in `.eslintrc.js`: https://github.com/eslint/eslint/blob/cdd063c388bbfe1781d7a864a832f03a2c1cc277/.eslintrc.js#L65

mdjermanovic replied:
Good catch! Updated in https://github.com/eslint/eslint/pull/17336/commits/6f71c8efe1f66aff9ad43eaf135bd33b1e8f4be8

aladdin-add reviewed (approved):
LGTM,  just waiting @nzakas to take another look.

nzakas commented:
> That's another option, but since the new config system doesn't have a feature like eslintrc's overrides.extends that handles matching multiple lists, I think it should be best practice to avoid files in shareable configs

In general, I agree that a best practice would be to avoid `files` for any shareable configs intended to be modified and extended by other people. In our case, as we don't intend `eslint-config-eslint` to be used by anyone else, I think we should make it be as drop-in as possible. The fewer steps we need to make to properly use the config, the less likely we are to make mistakes.

I'm also fine with putting a big warning message on the README to this effect. :)

mdjermanovic commented:
All right, I'll update the config to include `files`.

mdjermanovic commented:
Ready for another review.

aladdin-add reviewed (approved):
LGTM. 👍

nzakas reviewed (approved):
LGTM. Thanks!
